File: selenium-test-suite/tests/sdk-audit-graphics-multi-state.test.js
/**
 * Graphics Audit Tests with SDK (Selenium)
 *
 * Mirrors the Playwright graphics-audit.spec.js using Selenium WebDriver.
 */

const { By, until } = require('selenium-webdriver');
const { AccessFlowSDK, SeleniumDriver } = require('@acsbe/accessflow-sdk');
const { createDriver, navigateTo } = require('../helpers/driver');

AccessFlowSDK.init({
  apiKey:
    process.env.AF_NODE_PACKAGE_KEY ||
    process.env.AF_Node_Package_Key ||
    'flow-1TmAHDBGjDGX4aadlA0000VdvvxVnhZfP2',
});

let driver;

beforeAll(async () => {
  driver = await createDriver();
});

afterAll(async () => {
  if (driver) await driver.quit();
});

describe('Graphics Audit Tests with SDK', () => {
  beforeEach(async () => {
    await navigateTo(driver, '/');
    await driver.sleep(1000);
  });

  test('should perform audits across multiple states on background images page', async () => {
    const auditResults = [];
    const performAudit = async (context, options = {}) => {
      const { maxAttempts = 1, waitBetweenAttemptsMs = 2000 } = options;
      console.log(`Starting audit for context: ${context}`);
      let report = null;

      for (let attempt = 1; attempt <= maxAttempts; attempt++) {
        const sdk = new AccessFlowSDK(new SeleniumDriver(driver));
        report = await sdk.audit();
        if (report) break;

        if (attempt < maxAttempts) {
          console.log(
            `Audit returned no report for ${context} (attempt ${attempt}/${maxAttempts}), retrying...`,
          );
          await driver.sleep(waitBetweenAttemptsMs);
        }
      }

      auditResults.push({ context, report });
      if (!report) {
        console.log(`Audit completed for ${context} (no report after retries)`);
        return;
      }

      expect(report).toBeTruthy();
      console.log(`Audit completed for ${context}`);
    };

    const performSuccessPageAudit = async () => {
      await performAudit('Success Page', {
        maxAttempts: 4,
        waitBetweenAttemptsMs: 2500,
      });

      if (!auditResults[auditResults.length - 1]?.report) {
        // If the first pass still returns null, wait for page stability and retry once more.
        await driver.wait(
          until.elementLocated(By.css('main, [role="main"], h1, h2')),
          10000,
        );
        await driver.sleep(1200);
        await performAudit('Success Page (stability retry)', {
          maxAttempts: 2,
          waitBetweenAttemptsMs: 3000,
        });
      }
    };

    // Navigate to failure page
    await navigateTo(driver, '/graphics/background-images_failure');
    await driver.sleep(1000);
    const currentUrl = await driver.getCurrentUrl();
    expect(currentUrl).toContain('graphics/background-images_failure');

    // State 1: Initial Load (Failure Page)
    await performAudit('Initial Failure Page');

    // State 2: Open Modal (Notifications)
    try {
      const notificationButton = await driver.findElement(
        By.css('button:has(svg[data-testid="NotificationsIcon"])'),
      );
      if (await notificationButton.isDisplayed()) {
        await notificationButton.click();
        await driver.sleep(500);
        await performAudit('Notifications Modal Open');

        const body = await driver.findElement(By.css('body'));
        await body.sendKeys('\uE00C'); // Escape key
        await driver.sleep(500);
      }
    } catch {
      console.log('Notification button not found, skipping modal audit');
    }

    // State 3: Expand/Collapse Sidebar
    try {
      const closeSidebarButton = await driver.findElement(
        By.css('button:has(svg[data-testid="ChevronLeftIcon"])'),
      );
      if (await closeSidebarButton.isDisplayed()) {
        await closeSidebarButton.click();
        await driver.sleep(500);
        await performAudit('Sidebar Collapsed');

        try {
          const openSidebarButton = await driver.findElement(
            By.css('button:has(svg[data-testid="MenuIcon"])'),
          );
          if (await openSidebarButton.isDisplayed()) {
            await openSidebarButton.click();
            await driver.sleep(500);
            await performAudit('Sidebar Expanded');
          }
        } catch {
          // Menu button not found
        }
      }
    } catch {
      console.log('Sidebar toggle not found, skipping sidebar audit');
    }

    // State 4: Navigate to Success Page — wait for content then audit (retry up to 2x if null)
    await navigateTo(driver, '/graphics/background-images_success');
    await driver.sleep(2000);
    // Wait for success page content to be present
    await driver.wait(
      until.elementLocated(By.css('main, [role="main"], h1, h2')),
      10000,
    );
    await driver.sleep(800);

    await performSuccessPageAudit();

    // Verification — require at least 2 successful audits (Success Page can be flaky)
    console.log('Total audits performed:', auditResults.length);
    expect(auditResults.length).toBeGreaterThan(1);

    const successfulAudits = auditResults.filter((r) => r.report);
    expect(successfulAudits.length).toBeGreaterThanOrEqual(2);

    const violationCounts = auditResults.map((r) => ({
      context: r.context,
      violations: r.report?.violations ? r.report.violations.length : 0,
    }));
    console.log(
      'Violation counts by context:',
      JSON.stringify(violationCounts, null, 2),
    );
  });
});

File: java-selenium-tests/src/test/java/com/accessflow/SdkAuditGraphicsMultiStateTest.java
package com.accessflow;

import com.acsbe.accessflow.AccessFlowSDK;
import com.acsbe.accessflow.AccessFlowTeardown;
import com.acsbe.accessflow.SeleniumDriver;
import org.junit.jupiter.api.*;
import org.openqa.selenium.*;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.openqa.selenium.support.ui.WebDriverWait;

import java.time.Duration;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.*;

/**
 * Graphics Multi-State Audit Tests (Selenium).
 * Mirrors Node sdk-audit-graphics-multi-state.test.js.
 */
public class SdkAuditGraphicsMultiStateTest {

    private WebDriver driver;

    @BeforeEach
    void setup() {
        driver = DriverHelper.createDriver();
    }

    @AfterEach
    void teardown() {
        if (driver != null) driver.quit();
    }

    @AfterAll
    static void finalizeReports() {
        LocalReportCollector.write();
        AccessFlowTeardown.finalizeReports();
    }

    private AccessFlowSDK createSdk() {
        String apiKey = DriverHelper.getApiKey();
        return apiKey != null && !apiKey.isEmpty()
                ? new AccessFlowSDK(new SeleniumDriver(driver), apiKey)
                : new AccessFlowSDK(new SeleniumDriver(driver));
    }

    @Test
    @DisplayName("Should perform audits across multiple states on background images page")
    void testAuditAcrossMultipleStates() {
        List<Map<String, Object>> auditResults = new ArrayList<>();

        DriverHelper.navigateTo(driver, "/graphics/background-images_failure");
        DriverHelper.sleep(1000);
        assertTrue(driver.getCurrentUrl().contains("graphics/background-images_failure"));

        AccessFlowSDK sdk = createSdk();

        Map<String, Object> failureReport = sdk.audit();
        auditResults.add(Map.of("context", "Initial Failure Page", "report", failureReport != null ? failureReport : Map.of()));
        AccessFlowTeardown.recordAudit(driver.getCurrentUrl(), failureReport);
        LocalReportCollector.add(driver.getCurrentUrl(), failureReport);

        try {
            WebElement notificationBtn = driver.findElement(
                    By.xpath("//button[.//svg[@data-testid='NotificationsIcon']]"));
            if (notificationBtn.isDisplayed()) {
                notificationBtn.click();
                DriverHelper.sleep(500);
                Map<String, Object> modalReport = sdk.audit();
                auditResults.add(Map.of("context", "Notifications Modal Open", "report", modalReport != null ? modalReport : Map.of()));
                AccessFlowTeardown.recordAudit(driver.getCurrentUrl(), modalReport);
                LocalReportCollector.add(driver.getCurrentUrl(), modalReport);
                driver.findElement(By.cssSelector("body")).sendKeys(Keys.ESCAPE);
                DriverHelper.sleep(500);
            }
        } catch (NoSuchElementException ignored) {}

        try {
            WebElement closeSidebar = driver.findElement(
                    By.xpath("//button[.//svg[@data-testid='ChevronLeftIcon']]"));
            if (closeSidebar.isDisplayed()) {
                closeSidebar.click();
                DriverHelper.sleep(500);
                Map<String, Object> collapsedReport = sdk.audit();
                auditResults.add(Map.of("context", "Sidebar Collapsed", "report", collapsedReport != null ? collapsedReport : Map.of()));
                AccessFlowTeardown.recordAudit(driver.getCurrentUrl(), collapsedReport);
                LocalReportCollector.add(driver.getCurrentUrl(), collapsedReport);
                try {
                    WebElement openSidebar = driver.findElement(
                            By.xpath("//button[.//svg[@data-testid='MenuIcon']]"));
                    if (openSidebar.isDisplayed()) openSidebar.click();
                } catch (NoSuchElementException ignored) {}
            }
        } catch (NoSuchElementException ignored) {}

        DriverHelper.navigateTo(driver, "/graphics/background-images_success");
        DriverHelper.sleep(1000);
        assertTrue(driver.getCurrentUrl().contains("graphics/background-images_success"));

        // Retry for success page stability
        Map<String, Object> successReport = null;
        for (int i = 0; i < 4; i++) {
            successReport = sdk.audit();
            if (successReport != null) break;
            DriverHelper.sleep(2500);
        }
        if (successReport == null) {
            new WebDriverWait(driver, Duration.ofSeconds(10))
                    .until(ExpectedConditions.presenceOfElementLocated(By.cssSelector("main, [role='main'], h1, h2")));
            DriverHelper.sleep(1200);
            successReport = sdk.audit();
        }
        auditResults.add(Map.of("context", "Success Page", "report", successReport != null ? successReport : Map.of()));
        AccessFlowTeardown.recordAudit(driver.getCurrentUrl(), successReport);
        LocalReportCollector.add(driver.getCurrentUrl(), successReport);

        assertTrue(auditResults.size() > 1, "Should have at least 2 audit results");
    }
}
